xdr3/decode.go

```diff
@@ -95,6 +139,25 @@ type Decoder struct {
 	// used to minimize heap allocations during decoding
 	scratchBuf [8]byte
 	r          io.Reader
+	l          lenLeft
+}
+
+// NewDecoder returns a Decoder that can be used to manually decode XDR data
+// from a provided reader.  Typically, Unmarshal should be used instead of
+// manually creating a Decoder.
+func NewDecoder(r io.Reader) *Decoder {
+	if l, ok := r.(LenReader); ok {
+		return NewDecoderFromLenReader(l)
+	}
+	return &Decoder{r: r}
+}
+
+// NewDecoderFromLenReader works like NewDecoder, except it
+// uses the remaining input length provided by the len reader
+// to protect against doctored input XDR which could cause
+// allocation explosions.
+func NewDecoderFromLenReader(r LenReader) *Decoder {
```

Comment:
It seems to me that we now have two configuration options available to us when decoding:

1. max depth which is passed into `DecodeWithMaxDepth()`
2. max input length which is passed into `NewLenReader()`

I think it would be better to make these configuration options more explicit:

```go
type DecodeOptions struct {
	MaxDepth      int
	MaxInputLen  int
}

func NewDecoder(r io.Reader) *Decoder {
	return NewDecoderWithOptions(r, DecodeOptions{MaxDepth: DecodeDefaultMaxDepth, MaxInputLen: math.MaxInt32})
}

func NewDecoderWithOptions(r io.Reader, options DecodeOptions) *Decoder {
	...
}
```

Now that a decoder is always constructed with a `MaxInputLen`, you can remove the following code which follows `maxSize = d.mergeInputLenAndMaxSize(maxSize)` :

```go
	if maxSize == 0 {
		maxSize = maxInt32
	}
```

Reply:
Fwiw we went with something like this design in the Rust XDR, where we called it "Limits" where you provide a set of limits and that object gets passed around.

I considered "Options", and it seemed to work okay with the depth since when it was modified it was always unmodified on the way back up so options were consistent in each function, but it seemed odd that the options would be modified permanently with the length field, which is why I went with "Limits".

Reply:
I will implement @tamirms 's suggestion but I think I will stick with Options.